pkg/cmd/cli/cmd/create/clusterquota.go

```diff
@@ -0,0 +1,147 @@
+
+	clusterQuotaExample = `  # Create an cluster resource quota limited to 10 pods
+  %[1]s limit-bob --project-selector=openshift.io/requester=user-bob --hard=pods=10`
+)
+
+type CreateClusterQuotaOptions struct {
+	ClusterQuota *quotaapi.ClusterResourceQuota
+	Client       client.ClusterResourceQuotasInterface
+
+	Mapper       meta.RESTMapper
+	OutputFormat string
+	Out          io.Writer
+	Printer      ObjectPrinter
+}
+
+// NewCmdCreateServiceAccount is a macro command to create a new service account
+func NewCmdCreateClusterQuota(name, fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	o := &CreateClusterQuotaOptions{Out: out}
+
+	cmd := &cobra.Command{
+		Use:     name + " NAME --project-selector=key=value [--hard=RESOURCE=QUANTITY]...",
+		Short:   "Create cluster resource quota resource.",
+		Long:    clusterQuotaLong,
+		Example: fmt.Sprintf(clusterQuotaExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(o.Complete(cmd, f, args))
+			cmdutil.CheckErr(o.Validate())
+			cmdutil.CheckErr(o.Run())
+		},
+		Aliases: []string{"clusterquota"},
+	}
+
+	cmd.Flags().String("project-selector", "", "The project selector for the cluster resource quota")
+	cmd.MarkFlagRequired("project-selector")
+	cmd.Flags().StringSlice("hard", []string{}, "The resource to constrain: RESOURCE=QUANTITY (pods=10)")
+	cmdutil.AddOutputFlagsForMutation(cmd)
+	return cmd
+}
+
+func (o *CreateClusterQuotaOptions) Complete(cmd *cobra.Command, f *clientcmd.Factory, args []string) error {
+	if len(args) != 1 {
+		return fmt.Errorf("NAME is required: %v", args)
+	}
+
+	selector, err := unversioned.ParseToLabelSelector(cmdutil.GetFlagString(cmd, "project-selector"))
+	if err != nil {
+		return err
+	}
+
+	o.ClusterQuota = &quotaapi.ClusterResourceQuota{
+		ObjectMeta: kapi.ObjectMeta{Name: args[0]},
+		Spec: quotaapi.ClusterResourceQuotaSpec{
+			Selector: selector,
+			Quota: kapi.ResourceQuotaSpec{
+				Hard: kapi.ResourceList{},
+			},
+		},
+	}
+
+	for _, resourceCount := range cmdutil.GetFlagStringSlice(cmd, "hard") {
+		tokens := strings.Split(resourceCount, "=")
+		if len(tokens) != 2 {
+			return fmt.Errorf("%v must in the form of resource=quantity", resourceCount)
+		}
+		quantity, err := resource.ParseQuantity(tokens[1])
+		if err != nil {
+			return err
+		}
+		o.ClusterQuota.Spec.Quota.Hard[kapi.ResourceName(tokens[0])] = *quantity
+	}
+
+	o.Client, _, err = f.Clients()
+	if err != nil {
+		return err
+	}
+
+	o.Mapper, _ = f.Object(false)
+	o.OutputFormat = cmdutil.GetFlagString(cmd, "output")
+
+	o.Printer = func(obj runtime.Object, out io.Writer) error {
+		return f.PrintObject(cmd, o.Mapper, obj, out)
+	}
+
+	return nil
+}
+
+func (o *CreateClusterQuotaOptions) Validate() error {
+	if o.ClusterQuota == nil {
+		return fmt.Errorf("ClusterQuota is required")
+	}
+	if o.Client == nil {
+		return fmt.Errorf("Client is required")
```

Comment:
I would probably prefix with "internal error:" for start

On Wed, Jun 29, 2016 at 4:57 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/create/clusterquota.go
> <https://github.com/openshift/origin/pull/9588#discussion_r68961655>:
>
> > +	o.Mapper, _ = f.Object(false)
> > +	o.OutputFormat = cmdutil.GetFlagString(cmd, "output")
> > +
> > +	o.Printer = func(obj runtime.Object, out io.Writer) error {
> > +		return f.PrintObject(cmd, o.Mapper, obj, out)
> > +	}
> > +
> > +	return nil
> > +}
> > +
> > +func (o *CreateClusterQuotaOptions) Validate() error {
> > +	if o.ClusterQuota == nil {
> > +		return fmt.Errorf("ClusterQuota is required")
> > +	}
> > +	if o.Client == nil {
> > +		return fmt.Errorf("Client is required")
>
> The checks are fine, the error messages are not.
>
> the error messages can't assume these were fed via flags... what would you
> say instead?
>
> But what a user can do with those. If these will be nil it's a dev fault
> not a user/system error.
>
> devs are users, too :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9588/files/6549bac63cf40b3536c5756933aa7970efb598aa#r68961655>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADuFfyiMJxmb30gznfPJT87Fr-JdICdwks5qQoftgaJpZM4I_ZwA>
> .
>